[SelectionDAG] Preserve volatile undef stores.

enferex opened the pull request:
This patch preserves `undef` SDNodes that are `volatile` qualified. Previously, these nodes would be discarded.  The motivation behind this change is to adhere to the [LangRef](https://llvm.org/docs/LangRef.html#volatile-memory-accesses), even though that doc is mostly in terms of LLVM-IR, it seems reasonable to imply that the volatile constraints also imply to SDNodes.

> Certain memory accesses, such as [load](https://llvm.org/docs/LangRef.html#i-load)’s, [store](https://llvm.org/docs/LangRef.html#i-store)’s, and [llvm.memcpy](https://llvm.org/docs/LangRef.html#int-memcpy)’s may be marked volatile. The optimizers must not change the number of volatile operations or change their order of execution relative to other volatile operations. The optimizers may change the order of volatile operations relative to non-volatile operations. This is not Java’s “volatile” and has no cross-thread synchronization behavior.

Source: https://llvm.org/docs/LangRef.html#volatile-memory-accesses

enferex committed:
[SelectionDAG] Preserve volatile undef stores.

This patch preserves undef SDNodes that are volatile qualified.
Previously, these nodes would be discarded. The motivation behind this
change is to adhere to the LangRef, even though that doc is mostly in
terms of LLVM-IR, it seems reasonable to imply that the volatile
constraints also imply to SDNodes.

From the LangRef: https://llvm.org/docs/LangRef.html#volatile-memory-accesses
> Certain memory accesses, such as load's, store's, and llvm.memcpy's may
be marked volatile. The optimizers must not change the number of
volatile operations or change their order of execution relative to other
volatile operations. The optimizers may change the order of volatile
operations relative to non-volatile operations.

arsenm commented on llvm/test/CodeGen/AMDGPU/abi-attribute-hints-undefined-behavior.ll:
```diff
@@ -255,7 +255,7 @@ define amdgpu_kernel void @marked_kernel_use_other_sgpr(ptr addrspace(1) %ptr) #
   %implicitarg.load = load volatile i8, ptr addrspace(4) %implicitarg.ptr
   %dispatch.load = load volatile i8, ptr addrspace(4) %dispatch.ptr
-  store volatile i64 %dispatch.id, ptr addrspace(1) %ptr
+  store i64 %dispatch.id, ptr addrspace(1) %ptr
```
Why did this lose a volatile?

enferex replied:
By dropping the volatile we preserve the original FileCheck pattern.  I'm happy to restore the volatile and update the pattern match instead, that's probably the better choice given the intent of this patch.

arsenm replied:
It's supposed to just be a dummy use. You could equally just remove the store, I guess

arsenm commented on llvm/test/CodeGen/AMDGPU/bitcast-vector-extract.ll:
```diff
@@ -73,7 +73,7 @@ define amdgpu_kernel void @store_bitcast_constant_v8i32_to_v16i16(ptr addrspace(
   %undef = call i64 @llvm.amdgcn.icmp.i64(i64 %a, i64 %b, i32 999) #1
   %bc = bitcast i64 %undef to <2 x i32>
-  store volatile <2 x i32> %bc, ptr addrspace(1) %out
+  store <2 x i32> %bc, ptr addrspace(1) %out
```
Don't know why these tests are losing volatile

enferex replied:
In this case, the removal of the volatile preserves the original intent of the test, that there is no `store_dword`.  I'm not really sure what we should here since this test is explicitly checking for the absence of a store.

arsenm commented on llvm/test/CodeGen/NVPTX/store-undef.ll:
```diff
@@ -90,3 +90,19 @@ define void @test_store_def(i64 %param0, i32 %param1, ptr %out) {
+; CHECK-NEXT:    ret;
+  store volatile %struct.T undef, ptr %out
+  ret void
+}
```
Also test a vector, and store to poison

arsenm replied:
Specifically a vector that requires legalization splitting

arsenm replied:
No, I see those get turned into unreachable in opt -O3

gonzalobg replied:
And `atomic volatile` stores?

enferex committed:
Remove dummy stores from a few test cases and add a store-to-poison test.

The AMDGPU tests were cleaned up to remove the store instead of dropping
the volatile qualifier as I had done in the original patch.

The NVPTX test was expanded to test storing of a vector to poison.

arsenm commented on llvm/test/CodeGen/AMDGPU/mem-builtins.ll:
```diff
@@ -16,7 +16,6 @@ declare hidden i32 @strcmp(ptr nocapture, ptr nocapture) #1
 define amdgpu_kernel void @test_memcmp(ptr addrspace(1) %x, ptr addrspace(1) %y, ptr nocapture %p) #0 {
 entry:
   %cmp = tail call i32 @memcmp(ptr addrspace(1) %x, ptr addrspace(1) %y, i64 2)
-  store i32 %cmp, ptr addrspace(1) undef
```
Should preserve the dummy use and make volatile, or store to non-undef

arsenm commented on llvm/test/CodeGen/AMDGPU/mem-builtins.ll:
```diff
@@ -36,7 +34,6 @@ define amdgpu_kernel void @test_memchr(ptr addrspace(1) %src, i32 %char, i64 %le
 ; GCN: s_addc_u32 s{{[0-9]+}}, s{{[0-9]+}}, strcpy@rel32@hi+12
 define amdgpu_kernel void @test_strcpy(ptr %dst, ptr %src) #0 {
   %res = call ptr @strcpy(ptr %dst, ptr %src)
-  store ptr %res, ptr addrspace(1) undef
```
Really for all of these, should stop storing to undef. The point is to have a dummy use of the value, it doesn't matter what it is

arsenm commented on llvm/test/CodeGen/AMDGPU/lds-global-non-entry-func.ll:
```diff
@@ -153,7 +153,7 @@ define void @func_use_lds_global_constexpr_cast() {
 ; GISEL-NEXT:  .LBB1_2:
 ; GISEL-NEXT:    s_endpgm
-  store volatile i32 ptrtoint (ptr addrspace(3) @lds to i32), ptr addrspace(1) poison, align 4
+  store i32 ptrtoint (ptr addrspace(3) @lds to i32), ptr addrspace(1) poison, align 4
```
Keep volatile or make store to null or other filler pointer

arsenm commented on llvm/test/CodeGen/AMDGPU/mem-builtins.ll:
```diff
@@ -36,7 +34,6 @@ define amdgpu_kernel void @test_memchr(ptr addrspace(1) %src, i32 %char, i64 %le
 ; GCN: s_addc_u32 s{{[0-9]+}}, s{{[0-9]+}}, strcpy@rel32@hi+12
 define amdgpu_kernel void @test_strcpy(ptr %dst, ptr %src) #0 {
   %res = call ptr @strcpy(ptr %dst, ptr %src)
-  store volatile ptr %res, ptr addrspace(1) undef
```
Same keep all these dummy uses, avoid the undefs

enferex committed:
Update the AMDGPU/mem-builtins.ll to store to a non-dummy output formal.

enferex committed:
Update AMDGPU/lds-global-non-entry-func.ll: store to a dummy output formal.

arsenm commented on llvm/test/CodeGen/NVPTX/store-undef.ll:
```diff
@@ -90,3 +90,24 @@ define void @test_store_def(i64 %param0, i32 %param1, ptr %out) {
+; CHECK-NEXT:    st.volatile.v4.u32 [%rd4], {%r7, %r8, %r9, %r10};
+; CHECK-NEXT:    ret;
+  store volatile %struct.T undef, ptr %out
+  store volatile <8 x i32> %vec, <8 x i32>* poison
```
This is also store to poison, of poison. Should test both cases? Preferably in separate functions

enferex committed:
Create separate test routines to exercise testing poison.

enferex committed:
Fix a test variable name.

arsenm commented on llvm/test/CodeGen/NVPTX/store-undef.ll:
```diff
@@ -108,6 +108,40 @@ define void @test_store_volatile_undef(ptr %out, <8 x i32> %vec) {
 ; CHECK-NEXT:    ret;
   store volatile %struct.T undef, ptr %out
-  store volatile <8 x i32> %vec, <8 x i32>* poison
+  store volatile <8 x i32> %outVec, ptr undef
```
This is not testing the store of undef vector case. The vector is more relevant than the aggregate, which doesn't survive the DAG builder

enferex committed:
Test the volatile store of an undef vector.

RSilicon commented:
Should this be backpeoted to 19.x? Since something as big as volatile could have huge effects on codegen?

gonzalobg commented:
For the PTX backend at least, this is not critical. Whether a volatile write of undef is dropped or not is not observable by PTX applications.

gonzalobg commented on llvm/test/CodeGen/NVPTX/store-undef.ll:
```diff
@@ -90,3 +90,61 @@ define void @test_store_def(i64 %param0, i32 %param1, ptr %out) {
+; CHECK-NEXT:    ret;
+  store volatile %struct.T undef, ptr %out
+  store volatile <8 x i32> %vec, ptr undef
+  store volatile <8 x i32> undef, ptr %out
```
So just for my understanding, we are essentially lowering a volatile store of `undef`: 

```
store volatile i32 undef, ptr %out;
```

to 

```
{
   .reg .b32 %reg;
   st.volatile.u32 [ptr], %reg;
}
```

?

enferex replied:
@gonzalobg, yes.  In this case both will be lowered to `st.volatile.u32`

enferex commented:
> Should this be backpeoted to 19.x? Since something as big as volatile could have huge effects on codegen?

@RSilicon , this "issue" of the DAGCombiner dropping volatile undef stores has been around for quite a while, 10+years.  I'm not sure we want to back port this, I'm happy to do so, I do see one [bug ](https://github.com/llvm/llvm-project/issues/27683).  However, given it's been around for a while,  I can also see where it makes sense to just take it on as a fix for 20.
